hardware_interface/test/test_resource_manager.cpp

```diff
@@ -1329,3 +1329,238 @@ TEST_F(TestResourceManager, managing_controllers_reference_interfaces)
+      ASSERT_THAT(
+        failed_hardware_names,
+        testing::ElementsAreArray(std::vector<std::string>({TEST_SYSTEM_HARDWARE_NAME})));
+      status_map = rm.get_components_status();
+      EXPECT_EQ(
+        status_map[TEST_ACTUATOR_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+      EXPECT_EQ(
+        status_map[TEST_SYSTEM_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED);
+      interfaces_available(true, false);
+      rm.set_component_state(TEST_SYSTEM_HARDWARE_NAME, state_active);
+      status_map = rm.get_components_status();
+      EXPECT_EQ(
+        status_map[TEST_ACTUATOR_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+      EXPECT_EQ(
+        status_map[TEST_SYSTEM_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+      interfaces_available(true, true);
+    }
+    // write is sill OK
+    {
+      auto [ok, failed_hardware_names] = other_method(time, duration);
+      EXPECT_TRUE(ok);
+      EXPECT_TRUE(failed_hardware_names.empty());
+      interfaces_available(true, true);
+    }
+
+    // read failure for both, TEST_ACTUATOR_HARDWARE_NAME and TEST_SYSTEM_HARDWARE_NAME
+    claimed_itf_actuator.set_value(fail_value);
+    claimed_itf_system.set_value(fail_value);
+    {
+      auto [ok, failed_hardware_names] = method_that_fails(time, duration);
+      EXPECT_FALSE(ok);
+      EXPECT_FALSE(failed_hardware_names.empty());
+      ASSERT_THAT(
+        failed_hardware_names, testing::ElementsAreArray(std::vector<std::string>(
+                                 {TEST_ACTUATOR_HARDWARE_NAME, TEST_SYSTEM_HARDWARE_NAME})));
+      status_map = rm.get_components_status();
+      EXPECT_EQ(
+        status_map[TEST_ACTUATOR_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED);
+      EXPECT_EQ(
+        status_map[TEST_SYSTEM_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED);
+      interfaces_available(false, false);
+      rm.set_component_state(TEST_ACTUATOR_HARDWARE_NAME, state_active);
+      rm.set_component_state(TEST_SYSTEM_HARDWARE_NAME, state_active);
+      status_map = rm.get_components_status();
+      EXPECT_EQ(
+        status_map[TEST_ACTUATOR_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+      EXPECT_EQ(
+        status_map[TEST_SYSTEM_HARDWARE_NAME].state.id(),
+        lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE);
+      interfaces_available(true, true);
+    }
+    // write is sill OK
+    {
+      auto [ok, failed_hardware_names] = other_method(time, duration);
+      EXPECT_TRUE(ok);
+      EXPECT_TRUE(failed_hardware_names.empty());
+      interfaces_available(true, true);
+    }
+  };
+
+  using namespace std::placeholders;
+  // check read methods failures
+  check_read_or_write_failure(
```

Comment:
very neat!! I'd still like to refactor this a little bit to have 2 test cases, one to test if read fails, one to test if write fails and this implementation makes that pretty easy to do if you have a fixture for the rest

Reply:
Done